### PR TITLE
Feature: #102337 - PSR-14 event for modifying record list download pr…

### PR DESCRIPTION
**Pull request by linawolf**
…esets

References: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/934
Releases: main

### CODE REVIEW

File: Documentation/ApiOverview/Events/Events/Backend/BeforeRecordDownloadPresetsAreDisplayedEvent.rst [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+============================================
+
+..  versionadded:: 13.2
+
+The event :php:`TYPO3\CMS\Backend\RecordList\Event\BeforeRecordDownloadPresetsAreDisplayedEvent`
+can be used to manipulate the list of available download presets in
+the :guilabel:`Web > List` module.
+
+See :confval:`mod.web_list.downloadPresets <t3tsconfig:mod-web-list-downloadpresets>`
+On how to configure download presets.
```
Comment (froemken) [inline Documentation/ApiOverview/Events/Events/Backend/BeforeRecordDownloadPresetsAreDisplayedEvent.rst:16]:
```suggestion
on how to configure download presets.
```

File: Documentation/ApiOverview/Events/Events/Backend/_BeforeRecordDownloadPresetsAreDisplayedEvent/_PresetListener.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+declare(strict_types=1);
+
+namespace MyVendor\MyExtension\RecordList\EventListener;
+
+use TYPO3\CMS\Backend\RecordList\DownloadPreset;
+use TYPO3\CMS\Backend\RecordList\Event\BeforeRecordDownloadPresetsAreDisplayedEvent;
+use TYPO3\CMS\Core\Attribute\AsEventListener;
+
+#[AsEventListener(identifier: 'my-package/modify-record-list-preset')]
```
Comment (froemken) [inline Documentation/ApiOverview/Events/Events/Backend/_BeforeRecordDownloadPresetsAreDisplayedEvent/_PresetListener.php:11]: If we use `MyExtension` in namespace I would prefer using `my-extension` here. What do you think?

```suggestion
#[AsEventListener(identifier: 'my-extension/modify-record-list-preset')]
```

File: Documentation/ApiOverview/Events/Events/Backend/_BeforeRecordDownloadPresetsAreDisplayedEvent/_PresetListener.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            case 'tx_myvendor_myextension':
+                $presets[] = new DownloadPreset('PSR-14 preset', ['uid', 'something']);
+                break;
+        }
+
+        $presets[] = new DownloadPreset('Available everywhere, simple UID list', ['uid']);
+
+        $presets['some-identifier'] = new DownloadPreset('Overwrite preset', ['uid, pid'], 'some-identifier');
+
+        $event->setPresets($presets);
```
Comment (froemken) [inline Documentation/ApiOverview/Events/Events/Backend/_BeforeRecordDownloadPresetsAreDisplayedEvent/_PresetListener.php:37]: I also still use `switch`, but in times of PHP > 8.0 what do you think about a PHP:match solution?

```suggestion
    $presets = $event->getPresets();

    $newPresets = match ($event->getDatabaseTable()) {
        'be_users' => [new DownloadPreset('PSR-14 preset', ['uid', 'email'])],
        'pages' => [
            new DownloadPreset('PSR-14 preset', ['title']),
            new DownloadPreset('Another PSR-14 preset', ['title', 'doktype']),
        ],
        'tx_myvendor_myextension' => [new DownloadPreset('PSR-14 preset', ['uid', 'something'])],
    };

    foreach ($newPresets as $newPreset) {
        $presets[] = $newPreset;
    }

    $presets[] = new DownloadPreset('Available everywhere, simple UID list', ['uid']);
    $presets['some-identifier'] = new DownloadPreset('Overwrite preset', ['uid', 'pid'], 'some-identifier');

    $event->setPresets($presets);
```

**Review by froemken — COMMENTED** [review]
Thank you for your work. Found tiny improvements, if you want to adopt them.

File: Documentation/ApiOverview/Events/Events/Backend/_BeforeRecordDownloadPresetsAreDisplayedEvent/_PresetListener.php (hunk 3 above)
Comment (linawolf) [reply Documentation/ApiOverview/Events/Events/Backend/_BeforeRecordDownloadPresetsAreDisplayedEvent/_PresetListener.php:37]: Yeah, why not. I took the example as-is from the changelog but we can always improve it

**Review by froemken — DISMISSED** [review]

**Review by froemken — CHANGES_REQUESTED** [review]
Sorry. I was too fast. CGL fails

**Review by froemken — APPROVED** [review]
Thank you for catching the CGL problem.

**Pull request merged**